jobclient/python/jobs.py

```diff
@@ -0,0 +1,381 @@
+
+    @staticmethod
+    def from_string(name: str) -> 'Status':
+        """Parse a `Status` from a case-invariant string representation."""
+        return _JOB_STATUS_LOOKUP[name.lower()]
+
+
+_JOB_STATUS_LOOKUP = {
+    'initialized': Status.INITIALIZED,
+    'waiting': Status.WAITING,
+    'running': Status.RUNNING,
+    'completed': Status.COMPLETED
+}
+
+
+class State(Enum):
+    """The current state of a job.
+
+    Indicates whether a job is currently running, has finished and succeeded,
+    or has finished and failed.
+    """
+    WAITING = 'WAITING'
+    PASSED = 'PASSED'
+    FAILED = 'FAILED'
+
+    def __str__(self):
+        return self.value
+
+    def __repr__(self):
+        return f'State.{self.value}'
+
+    @staticmethod
+    def from_string(name: str) -> 'State':
+        """Parse a `State` from a case-invariant string representation."""
+        return _JOB_STATE_LOOKUP[name.lower()]
+
+
+_JOB_STATE_LOOKUP = {
+    'waiting': State.WAITING,
+    'passed': State.PASSED,
+    'failed': State.FAILED
+}
+
+
+class Application:
+    name: str
+    version: str
+
+    def __init__(self, name: str, version: str):
+        self.name = name
+        self.version = version
+
+    def to_dict(self) -> dict:
+        return {
+            'name': self.name,
+            'version': self.version
+        }
+
+    @classmethod
+    def from_dict(cls, d: dict) -> 'Application':
+        return cls(**d)
+
+
+class Job:
+    """A job object returned from Cook."""
+    command: str
+    mem: float
+    cpus: float
+
+    uuid: UUID
+    name: str
+    max_retries: int
+    max_runtime: int
+    status: Status
+    state: State
+    priority: int
+    disable_mea_culpa_retries: bool
+
+    executor: Optional[Executor]
+    expected_runtime: Optional[int]
+    pool: Optional[str]
+    instances: Optional[List[Instance]]
+    env: Optional[Dict[str, str]]
+    uris: Optional[List[FetchableUri]]
+    container: Optional[dict]
+    labels: Optional[Dict[str, str]]
+    constraints: Optional[Set[Constraint]]
+    groups: Optional[List[UUID]]
+    application: Optional[Application]
+    progress_output_file: Optional[str]
+    progress_regex_string: Optional[str]
+    user: Optional[str]
+    datasets: Optional[list]
+    gpus: Optional[float]
+    framework_id: Optional[str]
+    ports: Optional[int]
+    submit_time: Optional[int]
+    retries_remaining: Optional[int]
+
+    def __init__(self, *,
+                 # Required args
+                 command: str,
+                 mem: float,
+                 cpus: float,
+                 # Optional args with defaults
+                 uuid: Optional[UUID] = None,
+                 name: str = 'cookjob',
+                 max_retries: int = 5,
+                 max_runtime: int = sys.maxsize,
+                 state: State = State.WAITING,
+                 status: Status = Status.INITIALIZED,
+                 priority: int = 50,
+                 disable_mea_culpa_retries: bool = False,
```

Comment:
I don't think we should default these client-side. This is only for the query path, correct? I think we should just use what we get from the REST API.

Reply:
:+1: